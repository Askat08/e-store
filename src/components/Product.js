import React from "react";

const style = {
  img: {
    boxShadow:
      "0 4px 8px 0 rgba(0, 0, 0, 0.8), 0 6px 20px 0 rgba(0, 0, 0, 0.19)",
    cursor: "pointer",
  },
};

export default function Product(props) {
  const { title, path, ext, handleAddtoCart, id, price } = props;

  return (
    <div className="figure col-sm-4 col-md-2 p-0 my-3 mx-3">
      <img
        src={path + "." + ext}
        className="figure-img img-fluid rounded"
        alt={title}
        style={style.img}
      />

      <button
        className="btn-cart text-info shadow-none w-100"
        onClick={() => handleAddtoCart(id)}
      >
        <span className="mr-3">add to </span>
        <i className="fas fa-cart-plus" />
      </button>
      <p className="figure-caption mt-2 text-dark">
        <strong>{`${price}` <= 0 ? "free" : `$ ${price}`}</strong>
      </p>
      <p className="figure-footer text-dark">{title}</p>
    </div>
  );
}
