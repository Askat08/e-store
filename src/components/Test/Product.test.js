import React from "react";
import { shallow, mount } from "enzyme";
import Product from "../Product";
import Modal from "react-bootstrap/Modal";

const mockProps = {
  title: "comic",
  path: "http://path/to/image",
  ext: "jpg",
  id: 1,
  price: 12,
  description: "Superhero comic",
  cart: [],
  items: [],
  show: false,
};
const httpsMock = Array.from(mockProps.path);
// adding 's' to http on image url
httpsMock.splice(4, 0, "s");

describe("Product component test", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Product {...mockProps} />);
  });

  it("should render Product component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("button contains proper icon", () => {
    const button = wrapper.find("button");
    expect(button.contains(<i className="fas fa-cart-plus" />)).toBeTruthy();
  });

  it("paragraph has proper content", () => {
    const price = wrapper.find("p").at(0);
    const title = wrapper.find("p").at(1);
    expect(price.contains("$ " + mockProps.price)).toBeTruthy();
    expect(title.contains(mockProps.title)).toBeTruthy();
  });

  it("strong has proper content", () => {
    expect(wrapper.find("strong").text()).toEqual(
      mockProps.price <= 0 ? "free" : "$ " + mockProps.price
    );
  });

  it("should have proper props for image", () => {
    expect(wrapper.find("img[name='product-img']").props()).toEqual({
      name: "product-img",
      className: "figure-img img-fluid rounded",
      onClick: expect.any(Function),
      src: httpsMock.join("") + "." + mockProps.ext,
      alt: mockProps.title,
    });
    expect(wrapper.find("img").at(1).props()).toEqual({
      style: { width: "100%" },
      src: httpsMock.join("") + "." + mockProps.ext,
      alt: mockProps.title,
    });
  });

  it("button has proper props", () => {
    expect(
      wrapper.find("button").hasClass("btn-cart text-info shadow-none w-100")
    ).toBeTruthy();
    expect(wrapper.find("button").props().type).toEqual("button");
    expect(wrapper.find("button").props().onClick).toEqual(
      expect.any(Function)
    );
  });

  it("should test button onClick ", () => {
    const mockFunc = jest.fn();
    const wrapper = shallow(
      <Product handleAddtoCart={mockFunc} {...mockProps} />
    );

    wrapper.find("button[type='button']").invoke("onClick")();
    expect(mockFunc).toHaveBeenCalledWith(
      mockProps.id,
      mockProps.cart,
      mockProps.items
    );
  });

  it("should test handleShow and handleHide methods ", () => {
    wrapper.find("img[name='product-img']").invoke("onClick")();
    expect(wrapper.find(Modal).props().show).toBeTruthy();

    wrapper.find(Modal).invoke("onHide")();
    expect(wrapper.find(Modal).props().show).toBeFalsy();

    wrapper.find("img[name='product-img']").invoke("onClick")();
    expect(wrapper.find(Modal).props().show).toBeTruthy();
  });

  it("modal has proper props", () => {
    expect(wrapper.find(Modal.Header).hasClass("bg-warning")).toBeTruthy();
    expect(wrapper.find(Modal.Header).prop("closeButton")).toBeTruthy();
    expect(wrapper.find(Modal.Title).text()).toEqual(mockProps.title);
    expect(wrapper.find("p").at(2).hasClass("lead m-5 px-3")).toBeTruthy();
    expect(wrapper.find("small").hasClass("bg-warning")).toBeTruthy();
    expect(wrapper.find("small").text()).toEqual(mockProps.description);
  });
});
