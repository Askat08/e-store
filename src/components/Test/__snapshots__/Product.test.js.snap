// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product component test should render Product component 1`] = `
<div
  className="figure col-sm-4 col-md-3 col-lg-2 p-0 my-3 mx-3"
>
   
  <img
    alt="comic"
    className="figure-img img-fluid rounded"
    name="product-img"
    onClick={[Function]}
    src="https://path/to/image.jpg"
  />
  <button
    className="btn-cart text-info shadow-none w-100"
    onClick={[Function]}
    type="button"
  >
    <span
      className="mr-3"
    >
      add to 
    </span>
    <i
      className="fas fa-cart-plus"
    />
  </button>
  <p
    className="figure-caption mt-2 text-dark"
  >
    <strong>
      $ 12
    </strong>
  </p>
  <p
    className="figure-footer text-dark"
  >
    comic
  </p>
  <Bootstrap(Modal)
    className="text-dark"
    onHide={[Function]}
    show={false}
  >
    <div
      className="bg-info"
    >
       
      <Row
        noGutters={false}
      >
        <Col>
          <ModalBody>
            <img
              alt="comic"
              src="https://path/to/image.jpg"
              style={
                Object {
                  "width": "100%",
                }
              }
            />
          </ModalBody>
        </Col>
      </Row>
      <Row
        noGutters={false}
      >
         
        <Col>
          <ModalHeader
            className="bg-warning"
            closeButton={true}
            closeLabel="Close"
          >
            <ModalTitle>
              comic
            </ModalTitle>
          </ModalHeader>
          <p
            className="lead m-5 px-3"
          >
            <small
              className="bg-warning"
            >
              Superhero comic
            </small>
          </p>
        </Col>
      </Row>
    </div>
  </Bootstrap(Modal)>
</div>
`;
